File: haskell/src/Model/Changelog.hs
module Model.Changelog where

import Data.Aeson (ToJSON(..), object, (.=))
import Data.Text (Text)
import GHC.Generics (Generic)

data ChangelogChange = ChangelogChange
  { ccType        :: Text
  , ccDescription :: Text
  } deriving (Show, Generic)

instance ToJSON ChangelogChange where
  toJSON cc = object
    [ "type"        .= ccType cc
    , "description" .= ccDescription cc
    ]

data ChangelogItem = ChangelogItem
  { clVersion :: Text
  , clDate    :: Text
  , clChanges :: [ChangelogChange]
  } deriving (Show, Generic)

instance ToJSON ChangelogItem where
  toJSON cl = object
    [ "version" .= clVersion cl
    , "date"    .= clDate cl
    , "changes" .= clChanges cl
    ]

File: go/model/changelog.go
package model

// ChangelogChange represents a single change within a changelog entry.
type ChangelogChange struct {
	Type        string `json:"type"`
	Description string `json:"description"`
}

// ChangelogItem represents a single changelog entry containing multiple changes.
type ChangelogItem struct {
	Version string            `json:"version"`
	Date    string            `json:"date"`
	Changes []ChangelogChange `json:"changes"`
}
